File: practice/algorithms/implementation/utopian_tree/UtopianTree.scala
import scala.io.Source

object UtopianTree extends App {
  val nums = Source.stdin.getLines().toList.tail.map(_.toInt)
  nums.foreach(i => println(tree(0, i, 1)))

  def tree(actualCycle: Int, cycles: Int, height: Int): Int = {
    if(actualCycle == cycles) {
      return height
    }
    if(actualCycle % 2 == 0) tree(actualCycle + 1, cycles, height * 2)
    else tree(actualCycle + 1, cycles, height + 1)
  }
}

File: practice/algorithms/implementation/utopian_tree/UtopianTree.java
import java.util.Scanner;

public class UtopianTree
{
	public static void main(String[] args)
	{
		Scanner scanner = new Scanner(System.in);
		int testCases = scanner.nextInt();

		for (int i = 0; i < testCases; i++)
		{
			int cycles = scanner.nextInt();
			System.out.println(utopianTree(cycles));
		}

		scanner.close();
	}

	private static int utopianTree(int cycles)
	{
		int height = 1;
		for (int i = 0; i < cycles; i++)
		{
			if(i % 2 == 0)
			{
				height *= 2;
			}
			else
			{
				height += 1;
			}
		}
		return height;
	}
}
